### PR TITLE
HPCC-23360 DLLserver modified to use shared volume in cloud mode

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [x] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [x] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday For initial review

File: common/dllserver/dllserver.cpp [hunk 1]
```diff
@@ -759,37 +642,135 @@ void DllServer::removeDll(const char * name, bool removeDlls, bool removeDirecto
...
+    if (!target->exists())
+    {
+        source->copyTo(target, 0, NULL, true);
+    }
+
+    Owned<IRemoteConnection> conn = getEntryConnection(name, RTM_LOCK_WRITE);
+    if (conn)
+    {
+        /* check the entry doesn't exist already....
+         * Ideally the connection above would be a RTM_LOCK_READ and be changed to a RTM_LOCK_WRITE only when 'location' not found
```
Comment (ghalliday) [inline common/dllserver/dllserver.cpp]: That should be possible by calling conn->changeMode(), but are you likely to ever match an item that is alrady there?

File: dockerfiles/hpcc/templates/dllserver-claim.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: dllserver-pv-claim
+spec:
+  storageClassName: manual
+  accessModes:
+    - ReadWriteOnce
```
Comment (ghalliday) [inline dockerfiles/hpcc/templates/dllserver-claim.yaml]: Will this eventually change to ReadWriteMany?

File: dockerfiles/hpcc/templates/dllserver-claim.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  resources:
+    requests:
+      storage: 3Gi
```
Comment (ghalliday) [inline dockerfiles/hpcc/templates/dllserver-claim.yaml]: What is the long term plan for this?  Should this be a parameter configurable in the Values.yaml?

File: common/dllserver/dllserver.cpp (hunk 1 above)
Comment (richardkchapman) [reply common/dllserver/dllserver.cpp]: No. (This code is just copied from the legacy case)

File: dockerfiles/hpcc/templates/dllserver-claim.yaml (hunk 2 above)
Comment (richardkchapman) [reply dockerfiles/hpcc/templates/dllserver-claim.yaml]: Not sure what the difference is...

File: dockerfiles/hpcc/templates/dllserver-claim.yaml (hunk 3 above)
Comment (richardkchapman) [reply dockerfiles/hpcc/templates/dllserver-claim.yaml]: Yes, probably.

File: dockerfiles/platform-core/Dockerfile [hunk 4]
```diff
@@ -39,3 +39,4 @@ RUN dpkg -i *.deb ; \
 USER hpcc
 ENV PATH="/opt/HPCCSystems/bin:${PATH}"
 ENV HPCC_containerized=1
+ENV HPCC_DLLSERVER_PATH=/var/lib/HPCCSystems/queries
```
Comment (ghalliday) [inline dockerfiles/platform-core/Dockerfile]: Should there be and <eol> on the end of the file?

File: dockerfiles/hpcc/templates/dllserver-claim.yaml (hunk 2 above)
Comment (richardkchapman) [reply dockerfiles/hpcc/templates/dllserver-claim.yaml]: The access modes are: ReadWriteOnce – the volume can be mounted as read-write by a single node. ReadOnlyMany – the volume can be mounted read-only by many nodes. ReadWriteMany – the volume can be mounted as read-write by many nodes.

So yes, it should say ReadWriteMany if we want to go beyond a single node.

File: dockerfiles/hpcc/templates/eclagent.yaml [hunk 5]
```diff
@@ -12,8 +12,22 @@ spec:
       labels:
         run: eclagent
     spec:
+      volumes:
+      - name: dllserver-pv-storage
```
Comment (ghalliday) [inline dockerfiles/hpcc/templates/eclagent.yaml]: I imagine thor and roxie require similar treatment

File: dockerfiles/platform-core/Dockerfile (hunk 4 above)
Comment (richardkchapman) [reply dockerfiles/platform-core/Dockerfile]: Not sure what you mean

File: dockerfiles/hpcc/templates/eclagent.yaml (hunk 5 above)
Comment (richardkchapman) [reply dockerfiles/hpcc/templates/eclagent.yaml]: Yes, they will. Though Roxie may need some additional work as it uses a customized dllserver to handle disconnections from Dali

File: dockerfiles/platform-core/Dockerfile (hunk 4 above)
Comment (ghalliday) [reply dockerfiles/platform-core/Dockerfile]: The diff in github looks as if the last line isn't terminated with /n

**Review by ghalliday — COMMENTED** [review]
@richardkchapman I think this looks good to merge.  It needs squashing, and there are 3 files that are not terminated with \n, which would be cleaner if they were.

**Pull request merged**